Add support for external flash and sdram to imx rt series

MaureenHelm opened the pull request:
Adds support for the boot data, image vector table (IVT), and device configuration data (DCD) structures used by the imx boot ROM to boot an application from external xip flash and initialize the SDRAM. This increases the available flash and ram on the mimxrt1050_evk to 64 MB and 32 MB respectively.

MaureenHelm committed:
soc: nxp_imx: Add support for external xip flash boot header

Adds support for the boot data, image vector table, and FlexSPI NOR
config structures used by the imx rt boot ROM to boot an application
from an external xip flash device.

It is now possible to build and flash a bootable zephyr image to the
external xip flash on the mimxrt1020_evk, mimxrt1050_evk, and
mimxrt1060_evk boards via the 'ninja flash' build target and jlink
runner. Note, however, that the default board configurations still link
code into internal ITCM, therefore you must set CONFIG_CODE_HYPERFLASH=y
or CONFIG_CODE_QSPI=y explicitly to override the default. You must also
set CONFIG_NXP_IMX_RT_BOOT_HEADER=y to build the boot header into the
image.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
soc: boards: Reserve space for the imx boot header at the soc level

Moves the default TEXT_SECTION_OFFSET from the board level to the soc
level for the imx rt series. This offset is used to reserve space for
the imx boot header for external xip flash images.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: mimxrt1060_evk: Fix a Kconfig typo

Fixes a Kconfig typo in the mimxrt1060_evk board that was introduced
when the board was split into separate hyperflash and qspi
configurations.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
soc: nxp_imx: Add support for device configuration data (DCD)

Adds support for the device configuration data (DCD), which provides a
sequence of commands to the imx rt boot ROM to initialize components
such as an SDRAM.

It is now possible to use the external SDRAM instead of the internal
DTCM on the mimxrt1020_evk, mimxrt1050_evk, and mimxrt1060_evk. Note,
however, that the default board configurations still link data into
internal DTCM, therefore you must use a device tree overlay to override
"zephyr,sram = &sdram0"

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

jhqian reviewed (approved):
Tested boot from QSPI on rt1060 evk board and using SDRAM configured by DCD as well.

hakehuang commented:
@MaureenHelm , from segger official release only hyper flash and qspi flash are supported. Besides, there are some hard code change need to download to hyperflash

agansari reviewed (approved):
Changes look okay.

Suggestions:
1. Add memory increase to documentation, boards .yaml
2. Set external flash/ram as default?

MaureenHelm commented:
> Suggestions:
> 
> 1. Add memory increase to documentation, boards .yaml
> 2. Set external flash/ram as default?

Yes I plan to do both of these, but wanted to give you and others a chance to try it out before I change any defaults. Can you please check if there are any caching issues with Ethernet buffers in SDRAM?

galak commented on soc/arm/nxp_imx/rt/Kconfig.soc:
```diff
@@ -123,4 +123,41 @@ config IPG_DIV
 	int "IPG clock divider"
 	range 0 3
 
+menuconfig BOOT_HEADER
```
Should this be NXP_IMX_RT_BOOT_HEADER or something like that

MaureenHelm replied:
Done

galak reviewed (commented):
Generally looks good, only question was about the Kconfig symbol names.

MaureenHelm committed:
ext: mcux: Import SDK 2.5.0 evkmimxrt1020 board support

Imports board-level support for external xip flash and sdram from
MCUXpresso SDK 2.5.0.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
Updated to include rt1020 now that #12285 is merged
